src/include/OSL/oslnoise.h

```diff
@@ -253,437 +355,370 @@ inthash_simd (const int4& key_x, const int4& key_y, const int4& key_z, const int
     }
+
+    BaseNoiseT m_impl;
 };
 
+using PeriodicCellNoise = PeriodicAdaptionOf<CellNoise>;
+using PeriodicHashNoise = PeriodicAdaptionOf<HashNoise>;
+
+
 
 
 inline OSL_HOSTDEVICE int
 inthashi (int x)
 {
-    unsigned int i[1];
-    i[0] = (unsigned int)x;
-    return (int) inthash<1>(i);
+    return static_cast<int>(inthash(
+		static_cast<unsigned int>(x)
+	));
 }
 
 inline OSL_HOSTDEVICE int
 inthashf (float x)
 {
-    unsigned int i[1];
-    i[0] = OIIO::bit_cast<float,unsigned int>(x);
-    return (int) inthash<1>(i);
+    return static_cast<int>(
+		inthash(sfm::bit_cast<float,unsigned int>(x))
+	);
 }
 
 inline OSL_HOSTDEVICE int
 inthashf (float x, float y)
 {
-    unsigned int i[2];
-    i[0] = OIIO::bit_cast<float,unsigned int>(x);
-    i[1] = OIIO::bit_cast<float,unsigned int>(y);
-    return (int) inthash<2>(i);
+    return static_cast<int>(
+		inthash(
+			sfm::bit_cast<float,unsigned int>(x),
+			sfm::bit_cast<float,unsigned int>(y)
+		)
+	);
 }
 
 
 inline OSL_HOSTDEVICE int
 inthashf (const float *x)
 {
-    unsigned int i[3];
-    i[0] = OIIO::bit_cast<float,unsigned int>(x[0]);
-    i[1] = OIIO::bit_cast<float,unsigned int>(x[1]);
-    i[2] = OIIO::bit_cast<float,unsigned int>(x[2]);
-    return (int) inthash<3>(i);
+    return static_cast<int>(
+		inthash(
+			sfm::bit_cast<float,unsigned int>(x[0]),
+			sfm::bit_cast<float,unsigned int>(x[1]),
+			sfm::bit_cast<float,unsigned int>(x[2])
+		)
+	);
 }
 
 
 inline OSL_HOSTDEVICE int
 inthashf (const float *x, float y)
 {
-    unsigned int i[4];
-    i[0] = OIIO::bit_cast<float,unsigned int>(x[0]);
-    i[1] = OIIO::bit_cast<float,unsigned int>(x[1]);
-    i[2] = OIIO::bit_cast<float,unsigned int>(x[2]);
-    i[3] = OIIO::bit_cast<float,unsigned int>(y);
-    return (int) inthash<4>(i);
+    return static_cast<int>(
+		inthash(
+			sfm::bit_cast<float,unsigned int>(x[0]),
+			sfm::bit_cast<float,unsigned int>(x[1]),
+			sfm::bit_cast<float,unsigned int>(x[2]),
+			sfm::bit_cast<float,unsigned int>(y)
+		)
+	);
 }
 
 
+// Define select(bool,truevalue,falsevalue) template that works for a
+// variety of types that we can use for both scalars and vectors. Because ?:
+// won't work properly in template code with vector ops.
+// NOTE: Removing template and require explicit functions for different
+// combinations be specified using polymorphism.  Main reason is so that
+// different versions can pass parameters by value vs. reference.
+//template <typename B, typename F>
+//OSL_FORCEINLINE OSL_HOSTDEVICE F select (B b, const F& t, const F& f) { return b ? t : f; }
+
+OSL_FORCEINLINE OSL_HOSTDEVICE float select(const bool b, float t, float f) {
+    // NOTE:  parameters must NOT be references, to avoid inlining caller's creation of
+    // these values down inside the conditional assignments which can create to complex
+    // of a code flow for Clang's vectorizer to handle
+    static_assert(!std::is_reference<decltype(b)>::value, "parameters to select cannot be references");
```

Comment:
Aside: I think these static_assert calls are probably a holdover from when this function was a template. But now fully specialized to float, they don't serve a purpose. Some time later, we should just remove them.

Reply:
I think I left the static assert in there as strong teeth vs. just a comment, should someone come along later and change the parameters to references the static assert would fire.  In fact I might have done that myself and not wanted to repeat the same mistake.  It took a while to track it down, as inlining an entire AST tree inside a select because it is a reference is not obvious

Reply:
I'm a big fan of defensive programming. :-)